From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Thu, 18 Sep 2025 17:06:38 +0900
Subject: [PATCH] Reduce enderman teleport chunk lookups

Project: Airplane (https://github.com/TECHNOVE/Airplane)
Author: Paul Sauve <paul@technove.co>
License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

Project: Pufferfish (https://github.com/pufferfish-gg/Pufferfish)
Author: Kevin Raneri <kevin.raneri@gmail.com>
License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

Description:
* This patch is based on Pufferfish's "Reduce chunk loading & lookups" patch.

Airplane
Copyright (C) 2020 Technove LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of

diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index fcdb3ee493e65bf002e78862f00061336e731ede..a60ed071ab4c7c55e8cf1504ae96c917872c33d9 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -329,12 +329,18 @@ public class EnderMan extends Monster implements NeutralMob {
 
     private boolean teleport(double x, double y, double z) {
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos(x, y, z);
+        // Plazma start - Reduce enderman teleport chunk lookups
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(mutableBlockPos);
 
-        while (mutableBlockPos.getY() > this.level().getMinY() && !this.level().getBlockState(mutableBlockPos).blocksMotion()) {
+        if (chunk == null) {
+            return false;
+        }
+        // Plazma end - Reduce enderman teleport chunk lookups
+        while (mutableBlockPos.getY() > this.level().getMinY() && !chunk.getBlockState(mutableBlockPos).blocksMotion()) { // Plazma - Reduce enderman teleport chunk lookups
             mutableBlockPos.move(Direction.DOWN);
         }
 
-        BlockState blockState = this.level().getBlockState(mutableBlockPos);
+        BlockState blockState = chunk.getBlockState(mutableBlockPos); // Plazma - Reduce enderman teleport chunk lookups
         boolean flag = blockState.blocksMotion();
         boolean isWater = blockState.getFluidState().is(FluidTags.WATER);
         if (flag && !isWater) {
